utils/logger.go

```diff
@@ -0,0 +1,109 @@
+// this is a new logger interface for mattermost
+
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"runtime"
+
+	l4g "github.com/alecthomas/log4go"
+)
+
+// this pattern allows us to "mock" the underlying l4g code when unit testing
+var debug = l4g.Debug
+var info = l4g.Info
+var err = l4g.Error
+
+// Logger - an instance of the log framework that can be used to log messages
+type Logger string
+
+// NewLogger - creates a new instance of Logger named after the file that created it
+func NewLogger() Logger {
+	_, file, _, ok := runtime.Caller(1)
+	if ok {
+		return Logger(file)
+	}
+	return Logger("Unknown Logger")
+}
+
+// contextKey lets us add contextual information to log messages
+type contextKey string
```

Comment:
This is the recommended  approach when dealing with Golang's Context object. It's an anti-pattern to use strings as keys on the Context, because those strings could collide with other keys that some other part of the codebase puts on the Context. Instead, it's recommended that each bit of code create its own types to use as keys so that two different parts of the codebase could put the same key on the Context without them colliding and overwriting one another.

For example, in this class, I'm creating context keys with the names "user-id" and "request-id," but it's not inconceivable that some other part of the codebase might want to create keys with the same names. By making the keys of type `contextKey`, it's impossible for my keys to ever collide with somebody else's keys, even if they have the same name.

Check out this blog post for more: https://blog.golang.org/context